state-chain/pallets/cf-ingress-egress/src/boost_pool/tests.rs

```diff
@@ -0,0 +1,427 @@
+		let expected_amounts = &expected_boosts[boost_id];
+
+		assert_eq!(
+			BTreeMap::from_iter(expected_amounts.iter().copied()),
+			BTreeMap::from_iter(
+				boost_amounts.iter().map(|(id, amount)| (*id, amount.into_chain_amount()))
+			)
+		)
+	}
+}
+
+#[track_caller]
+fn check_pending_withdrawals(
+	pool: &TestPool,
+	withdrawals: impl IntoIterator<Item = (AccountId, Vec<BoostId>)>,
+) {
+	let expected_withdrawals: BTreeMap<_, BTreeSet<_>> = withdrawals
+		.into_iter()
+		.map(|(account_id, boost_ids)| (account_id, boost_ids.into_iter().collect()))
+		.collect();
+
+	assert_eq!(pool.pending_withdrawals, expected_withdrawals, "mismatch in pending withdrawals");
+}
+
+#[test]
+fn test_scaled_amount() {
+	use cf_chains::Ethereum;
+	// This shows that we can have unreasonably large amounts in chains with
+	// a large number of decimals and still fit into u128 after scaling up:
+
+	// 1 trillion FLIP (or ETH; other chains have smaller number of decimals)
+	let original: EthAmount = 1_000_000_000_000 * FLIPPERINOS_PER_FLIP;
+	let scaled: ScaledAmount<Ethereum> = ScaledAmount::from_chain_amount(original);
+	let recovered: EthAmount = scaled.into_chain_amount();
+	assert_eq!(original, recovered);
+}
+
+#[test]
+fn adding_funds() {
+	let mut pool = TestPool::new(5);
+
+	pool.add_funds(BOOSTER_1, 1000);
+	check_pool(&pool, [(BOOSTER_1, 1000)]);
+
+	pool.add_funds(BOOSTER_1, 1000);
+	check_pool(&pool, [(BOOSTER_1, 2000)]);
+
+	pool.add_funds(BOOSTER_2, 1000);
+	check_pool(&pool, [(BOOSTER_1, 2000), (BOOSTER_2, 1000)]);
+}
+
+#[test]
+fn withdrawing_funds() {
+	let mut pool = TestPool::new(5);
+	pool.add_funds(BOOSTER_1, 1000);
+	pool.add_funds(BOOSTER_2, 1000);
+	pool.add_funds(BOOSTER_3, 1000);
+	check_pool(&pool, [(BOOSTER_1, 1000), (BOOSTER_2, 1000), (BOOSTER_3, 1000)]);
+
+	// No pending to receive, should be able to withdraw in full
+	assert_eq!(pool.stop_boosting(BOOSTER_1), Ok(1000));
+	check_pool(&pool, [(BOOSTER_2, 1000), (BOOSTER_3, 1000)]);
+	check_pending_withdrawals(&pool, []);
+
+	assert_eq!(pool.stop_boosting(BOOSTER_2), Ok(1000));
+	check_pool(&pool, [(BOOSTER_3, 1000)]);
+
+	assert_eq!(pool.stop_boosting(BOOSTER_3), Ok(1000));
+	check_pool(&pool, []);
+}
+
+#[test]
+fn withdrawing_twice_is_no_op() {
+	let mut pool = TestPool::new(0);
+	pool.add_funds(BOOSTER_1, 1000);
+	pool.add_funds(BOOSTER_2, 1000);
+
+	assert_eq!(pool.stop_boosting(BOOSTER_1), Ok(1000));
+
+	check_pool(&pool, [(BOOSTER_2, 1000)]);
+
+	assert!(pool.stop_boosting(BOOSTER_1).is_err());
+
+	// No changes:
+	check_pool(&pool, [(BOOSTER_2, 1000)]);
+}
+
+#[test]
+fn boosting_with_fees() {
+	let mut pool = TestPool::new(100);
+
+	pool.add_funds(BOOSTER_1, 1000);
+	pool.add_funds(BOOSTER_2, 2000);
+
+	check_pool(&pool, [(BOOSTER_1, 1000), (BOOSTER_2, 2000)]);
+
+	assert_eq!(pool.provide_funds_for_boosting(BOOST_1, 1010), Ok((1010, 10)));
+
+	// The recorded amounts include fees (1 is missing due to rounding errors in *test* code)
```

Comment:
The tests use boost pool's interface with AssetAmounts, but internally the amounts are represented more accurately. When the fees are distributed, we give 3333 "subatomic" units to one node and 6667 to the other. I had to make a somewhat difficult choice between asserting with atomic amounts in tests (which would be confusing since we use AssetAmounts in all interfaces), or to allow asserting on AssetAmounts, but have them be converted to subatomic amounts in helper functions. Reasonable arguments could be made in favor of either one, but in the end I settled on the lattter.

Now, because there is no way to specify a fractional AssetAmount (3.333), I have to convert one into the other, which can introduce rounding errors. In this case specifically I added a commet to address a potential confusion as to why 3 + 6 in fees don't add up to 10. I point out that it is an "error" in the test, because internally we still keep the fractional amounts. This actually makes sense since if boosters were to withdraw their funds immediately, they would be getting 3 and 6.